### ISSUE TITLE
Intelligently create temporary resource storage

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1446]
**Is your feature request related to a problem? Please describe.**
- Currently, if a resource that's produced exceed its initial capacity, it tosses away the excess amount.
- The Storage class allocates specific resource capacity for each resource in each building.
- The concept of General Cargo Storage is NOT well implemented.

**Describe the rationale**
- Allow the instant use of new Generic Container or the new Storage Bin to store the excess resources amount .

**Describe the goal(s) you have in mind**
- Put a new empty, Generic Container into use in order to temporary store the new resource just being produced.
- For long term storage, prompt players to build warehouse building that have designated storage for specific storage.
- Rework and refine the concept of General Cargo Storage

**Describe the non-goal(s) associated with the goal(s)**
- Currently, a mobile container only needs to be known to belong to a settlement. 
- There is no need to designate a building for where a **mobile** container is stored. 

**Stacktrace**
```
01-Adir-02:715.716 (Info) Manufacture [x19] : [Teslapolis] Workshop 1 - Finished the manu process 'Make Nitrites'.
01-Adir-02:715.716 (Severe) Manufacture : [Teslapolis] Workshop 1 - Not enough storage capacity to store 1.6505696686669014 of nitrite from Make Nitrates
...
01-Adir-02:862.161 (Severe) Manufacture : [Teslapolis] Workshop 1 - Not enough storage capacity to store 1.8500000000000059 of sodium hypochlorite from Make sodium hypochlorite for bleaching and sanitation
```

**Specifications**
- pre-3.9.0

**Comment by mokun** [comment]
@bevans2000 

How would you like to handle this ?

**Comment by bevans2000** [comment]
I can have a look towards the end of 3.9.0

**Comment by mokun** [comment]
@bevans2000 

I kind of update this thread to reflect the same situation I used to encounter but it never gets resolved.

The real problem is that when the message `Specific Storage is full...` or `Stock Storage is full...`, the simulation would nearly halt as logger gets called 40,000+ times.

My VisualVM registered a spike in CPU utilization when that happens.

So what's our approach in getting rid of the intensive log while at the same time what actions the settlement should get triggered to do to have more space for storing excess resources ?

```
01-Adir-18:197.846 (Warning) MicroInventory [x7370] : [72.7515 N 164.7992 E] Obruchev - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.875 (Warning) MicroInventory [x7372] : [6.4117 N 20.4320 W] Rebus - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.875 (Warning) MicroInventory [x7372] : [26.3959 N 71.0983 W] Vikram - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.875 (Warning) MicroInventory [x28912] : [40.0805 N 156.5333 W] Drifter - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.875 (Warning) MicroInventory [x7372] : [40.0546 N 156.5256 W] Serenity - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.875 (Warning) MicroInventory [x14279] : [6.0549 N 164.4946 E] So-Ki - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
01-Adir-18:197.889 (Warning) MicroInventory [x22121] : [17.5430 N 81.8992 E] Lomonosov - Stock Storage is full. Excess 0.0 kg Carbon dioxide.

```

Some resources such as CO2 are useless for sure. But regolith and others are useful.

**Comment by mokun** [comment]
Feels like if the settlement is careful enough to do an inventory of remaining capacity. 

Say when regolith reaches 99% capacity, it should start to alert the `Chief of Resource` and plan according on how to meet the demand for more specific storage space and more stock storage space.

It should trigger manufacturing containers such as gas canister or barrel to store the resources before they become full.

**Comment by mokun** [comment]
`01-Adir-18:197.846 (Warning) MicroInventory [x7370] : [72.7515 N 164.7992 E] Obruchev - Stock Storage is full. Excess 0.0 kg Carbon dioxide.
`

Interesting, this includes a vehicle's storage space from the log above. 

In reality, it should be happening since the gases happen. But we have no specific storage space for CO2 in a vehicle. 

And we did model the release of CO2 from a person in `PhysicalCondition`. 

So how should we UNIVERSALLY handle both excess gases and excess solid resources like Regolith ?

**Comment by mokun** [comment]
- Below it shows that on sol 18, Regolith already has 65k kg. Since its specific capacity is also at 65k. It's full. 
<img width="528" height="637" alt="Image" src="https://github.com/user-attachments/assets/1bda4db7-7231-47ef-be3d-1abe71ab7b7f" />



- On the other hand, below it shows the settlement Foundation has nearly 95k regolith. I wonder why the discrepancy. Is it 65k or 95k ?
<img width="1527" height="475" alt="Image" src="https://github.com/user-attachments/assets/8e2cd3da-98c3-470c-8b04-8273bdbf8dab" />

**Comment by bevans2000** [comment]
I think the logging class is not very good when it comes to collapsing repeated messages. I can take a look but in deep with the Construction changes at the moment.

**Comment by bevans2000** [comment]
I should have time free from work tonight to create an Issue outlining the suite of related changes

**Comment by bevans2000** [comment]
> > logging class is not very good when it comes to collapsing repeated messages
> 
> Yay there may be bugs in handling repeated messages 
> 
> And it's possible the java logging is not that good. 
> 
> Can we try using some other logging API out there that are less CPU intensive ?
> 
> 
Writing to screen is expensive and that will slow it down. But we can change the throttling approach by making better use of the Entity argument to find hits in the cache. I was thinking we could apply throttling to all messages to prevent repeated outputs.

**Comment by mokun** [comment]
@bevans2000 

So I imagine you can modify SimLogger to save or generate a report of how many similar log msg have been captured and sort them out.

The result would be accessible via the console/terminal.

I was also thinking about creating a console command to change how to handle spurious logs such as `Storage is full`.

**Comment by bevans2000** [comment]
> 
> I was also thinking about creating a console command to change how to handle spurious logs such as `Storage is full`.

I've tweaked the SimLogger already. There is a console command already to control logging at he class level.
The root cause here is logging of a lack of resources is outside the Resource holder in many places. Plus if have a check and extract pattern everywhere. What we need is a basic extract what is available pattern and then the log can be inside the holder. The benefit is we can use a latch pattern so only the first case of a resource storage is logged. The latch is reset when a resource is added.
We can improve this for sure but I want to avoid any temp. storage because that is just moving the problem.

**Comment by bevans2000** [comment]
The problem with the full storage is some parts of the simulator do not consider it there is capacity before starting. If the storage is full, it's full. We can't just magically create temporary storage.
I think a mitigation is to only log the message the first time it happens instead of every time.
